### ISSUE TITLE
Path to Model Weights

### ISSUE THREAD

**Issue by NikhilMank** [nish03/FFS#2]
I have ran the training of the network with VOC dataset as shown in the repo and ran the inference like shown. The results which I have got were not matching with that shown in the paper, they were very high (FPR95 was around 95). I thought there was some issues in the training. So, I tested again by replacing the model.pth with the pre-trained network which I have downloade from the link provided but still I got the same result as with the trained network. I don't know what is going wrong here. 

This is the output which I got when I have tested with the pre-trained network:
![image](https://github.com/nish03/FFS/assets/41358516/c199c393-8e8c-403a-8d00-8389c08fe7d8)

the ouput after inference with coco_ood_val:
![image](https://github.com/nish03/FFS/assets/41358516/68907975-c747-444c-8caf-a2affdbc0968)

the output after inference with voc_val:
![image](https://github.com/nish03/FFS/assets/41358516/d68ff70c-9b13-4a04-8358-13f1247135f2)

Please let me know if any additional info is needed, I have given the information needed to the best of my knowledge.

**Comment by nish03** [comment]
Hi, thanks for the query. Your energy threshold is 0.0906, which you should not get. This problem occurs when the inference code uses the weights from `regnetx_detectron2.pth` in the path `Flow-Feature-Synthesis/detection/configs`. For your information, `regnetx_detectron2.pth` contains the weights of the pre-trained backbone network. In contrast, your inference code should use the pre-trained FFS model weights provided in the link on README page. Please check the README section titled 'Using Pre-trained FFS Models' to know how to place the pre-trained FFS weights in the correct folder path for the inference code to use these weights rather than `regnetx_detectron2.pth`. Once you use the pre-trained FFS weights, the inference should first be made on inlier PASCAL-VOC and only then on the outlier dataset such as MS-COCO. When the inference on PASCAL-VOC is complete, the energy threshold should be 0.5959, as also mentioned on the README page. I hope this answers your question.

**Comment by NikhilMank** [comment]
Hi, Thanks for your quick response. The threshold 0.0906 was obtained when I have trained the network with the config file provided. I don't know why the threshold was that off. As per my understanding, all the required inputs are in the regnetx.yaml file and that should reproduce somewhat close results around the published value. Is that file the only thing or are there anything else to be observed during training. Also, I have used only one GPU instead of 4 ( I don't think that will affect the process).

I have downloaded the final_model.pth from the provided drive link and placed it at `/home/jovyan/FFS/Flow-Feature-Synthesis/detection/data/VOC-Detection/faster-rcnn/regnetx/random_seed_0/` as mentioned in the first two steps of using pre-trained model section. Later I have perfomed the inference procedure like mentioned. It is still generating the same threshold value.

I have deleted all the previous .pkl, .pth, .txt files, deleted the `final_model.pth` and `regnetx_detectron2.pth`, redownloaded it from the links in the README.md, placed them at the respective places and then ran the inference on custom_voc_val and coco_ood_val before typing this comment but still I got the same result for threshold. 

The location of the model:
![image](https://github.com/nish03/FFS/assets/41358516/750ef000-33bb-497e-9d5d-838caa10d317)

The location of the regnetx:
![image](https://github.com/nish03/FFS/assets/41358516/2d6ebbd4-6512-4928-88bd-8ba956b96e62)

Results of voc_custom_val:
![image](https://github.com/nish03/FFS/assets/41358516/f83a0867-ea78-4522-bb21-b3c26b728591)

Results of coco_ood_val:
![image](https://github.com/nish03/FFS/assets/41358516/28bb0383-ff98-4712-a239-3cb25eeab559)

The config file:
![image](https://github.com/nish03/FFS/assets/41358516/e9b7b63c-84c6-485c-b722-8b21e5beb343)

**Comment by nish03** [comment]
In the config file, could you change the WEIGHTS path to `/path/to/Flow-Feature-Synthesis/detection/data/VOC-Detection/faster-rcnn/regnetx/random_seed_0/model_final.pth` and then again rerun the inference for just PASCAL-VOC?

**Comment by NikhilMank** [comment]
I have changed the path and then ran the inference using `python apply_net.py  --dataset-dir /home/jovyan/Datasets/VOC_DATASET_ROOT/ --test-dataset voc_custom_val --config-file VOC-Detection/faster-rcnn/regnetx.yaml --inference-config Inference/standard_nms.yaml --random-seed 0 --image-corruption-level 0 --visualize 0` but still the threshold value is 0.0906

![image](https://github.com/nish03/FFS/assets/41358516/8d73df4d-b7a9-4bf4-bd9f-4fa9cb03d8c4)
![image](https://github.com/nish03/FFS/assets/41358516/aca36cf0-9074-4481-9e79-2965d9cc1b54)

**Comment by NikhilMank** [comment]
Hi, I have changed the path in the config file but it didn't change the threshold value. let me know if you need anything else for better understanding of the issue. like `config.yaml` `regnetx.yaml` `log.txt` files produced during inference or any other information to shed more light.

Thank you.

**Comment by nish03** [comment]
Hi @NikhilMank, It is weird that you still get the threshold value of 0.0906, which means the model is not trained at all and just uses the initial weights. I and a couple of other users had reproduced the results using the same weights provided on the drive which means there is a bug in your code version setup which is hard to know. What I can do is I will run the code again on my machine with the weights I shared on the drive and try to reproduce the results. I will get back to you as soon as possible.

**Comment by NikhilMank** [comment]
Hi, I will also delete everything, re-download everything again, and do it fresh. I will let you know once after that. 

Thank you for your time, much appreciated.

**Comment by NikhilMank** [comment]
Hi, Now doing it again fresh made me realised I made a few changes in the inference. I was getting not on same device error when I ran the inference code so I have did the following changes as the errors were shown in the below order:

- when I ran inference I got not on same device error @line157 in detection/inference/rcnn_predictor.py so I changed the line to  "det_labels = det_labels[filter_mask.to(det_labels.device)]" from "det_labels = det_labels[filter_mask]
"
- same happened @ line84 of detection/inference/inference_utils.py so I changed the line to "result.det_labels = det_labels[keep.to(det_labels.device)]"  from "result.det_labels = det_labels[keep]"
- same happened @line463 of the same above file so I have replaced the line with “results = results[output_boxes.nonempty().to(results.pred_boxes.tensor.device).cpu()]”

then the inference could continue without error but the threshold is the same. I don't know if these changes are the reason for the wrong inference or not but these are the changes I made. The other code is exactly the same as in your repo.

**Comment by NikhilMank** [comment]
Hey, I think I know the issue now. I did it again fresh and this time I got a threshold of 0.5971 for voc_custom_val.

The issue was there was a checkpoint saved, the checkpoint was regnetx_detectron2.pth, and it was loading that model despite of giving the model weights in the config file. I don't know how to solve it yet but that was the issue.

This is how it should have happen 
![image](https://github.com/nish03/FFS/assets/41358516/fc23023a-d41a-4bc6-9b19-51a9fde1fd85)

but it was happening like this
![image](https://github.com/nish03/FFS/assets/41358516/c3404064-96ad-4943-b80e-abc1cd839d9a)


- Do you have any suggestion on how can I solve this in my old workspace? 
- Also, Please add that we have to change the path to weights in the README.md file as students who are beginners like me don't know how much about operating through config files.

Thank you.

**Comment by nish03** [comment]
Hi, great that the issue is solved according to my suggestion. Yes, I will add a short note in the ReadME file about the correct way of loading the weights when using the pre-trained model.

**Comment by NikhilMank** [comment]
Hi, I did not understand why it is searching for a checkpoint during inference. It made sense for me doing that during the training but I don't understand why we need that during infernece? Can you also please tell me how to prevent this from hapenning, like which part should I change in the code base?

Thank you.

**Comment by nish03** [comment]
Hi @NikhilMank , you cannot prevent this as I think it is a functionality of the Detectron2 framework itself. Please check the class `fvcore.common.checkpoint.Checkpointer` in this [link](https://detectron2.readthedocs.io/en/latest/modules/fvcore.html#module-fvcore.common.checkpoint). It clearly describes that in addition to the trained model weights, the functionality will also search for any checkpointable objects in the code setup. It is not a problem during the training phase since the code setup just has a single model checkpoint in the form of `regnetx_detectron2.pth`. However, during inference, there are two checkpoints in the code setup in the form of `model_final.pth` and `regnetx_detectron2.pth`. Hence, `fvcore.common.checkpoint.Checkpointer` uses `regnetx_detectron2.pth` as per the config file assignment. Hope this helps.
